opencensus/common/schedule/__init__.py

```diff
@@ -50,3 +52,121 @@ def run(self):
 
     def cancel(self):
         self.finished.set()
+
+
+class QueueEvent(object):
+    def __init__(self, name):
+        self.name = name
+        self.event = threading.Event()
+
+    def __repr__(self):
+        return ('{}({})'.format(type(self).__name__, self.name))
+
+    def set(self):
+        return self.event.set()
+
+    def wait(self, timeout=None):
+        return self.event.wait(timeout)
+
+
+class QueueExitEvent(QueueEvent):
+    pass
+
+
+class Queue(object):
+    def __init__(self, capacity):
+        self.EXIT_EVENT = QueueExitEvent('EXIT')
+        self._queue = queue.Queue(maxsize=capacity)
+
+    def _gets(self, count, timeout):
+        start_time = time.time()
+        elapsed_time = 0
+        cnt = 0
+        while cnt < count:
+            try:
+                item = self._queue.get(block=False)
+                yield item
+                if isinstance(item, QueueEvent):
+                    return
+            except queue.Empty:
+                break
+            cnt += 1
+        while cnt < count:
+            wait_time = max(timeout - elapsed_time, 0)
+            try:
+                item = self._queue.get(block=True, timeout=wait_time)
+                yield item
+                if isinstance(item, QueueEvent):
+                    return
+            except queue.Empty:
+                break
+            cnt += 1
+            elapsed_time = time.time() - start_time
+
+    def gets(self, count, timeout):
+        return tuple(self._gets(count, timeout))
+
+    def flush(self, timeout=None):
+        start_time = time.time()
+        wait_time = timeout
+        event = QueueEvent('SYNC(timeout={})'.format(wait_time))
+        try:
+            self._queue.put(event, block=True, timeout=wait_time)
+        except queue.Full:
+            return
+        elapsed_time = time.time() - start_time
+        wait_time = timeout and max(timeout - elapsed_time, 0)
+        if event.wait(timeout):
+            return time.time() - start_time  # time taken to flush
+
+    def put(self, item, block=True, timeout=None):
+        try:
+            self._queue.put(item, block, timeout)
+        except queue.Full:
+            pass  # TODO: log data loss
```

Comment:
This is a good use for metrics, we should try to emit the same metric for all clients for unexported spans.

Reply:
Yep!

Reply:
Here goes the metrics that we should consider https://github.com/census-instrumentation/opencensus-specs/issues/262#issuecomment-488881213.